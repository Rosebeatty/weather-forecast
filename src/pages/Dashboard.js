import React, { useEffect, useState } from "react";
import { getLocations } from "services/WeatherService";
import { CardList } from "components/CardList";
import { ToggleSwitch } from "components/Toggle";
import { LocationSearch } from "components/LocationSearch";
import { changeUnit } from "utils/changeUnit";
import styled from "styled-components";

import "react-toggle/style.css";

const Layout = styled.div`
  display: flex;
  height: 80vh;
  margin-top: 40px;
  flex-direction: column;
  justify-content: center;
  padding: 2em;
`;

const PopupContainer = styled.div`
  position: fixed:
  background: rgba(0, 0, 0, 0.5):
  height: 100vh;
  width: 100vw;
  top: 0px;
  margin-left: -22px;
`;

const Popup = styled.div`
  z-index: 10000;
  background-color: #212121;
  backdrop-filter: blur(20px);
  border: purple;
  position: fixed;
  top: 50%;
  left: 50%;
  width: 350px;
  padding: 3em 2em;
  border-radius: 10px;
  transform: translate(-50%, -50%);
`;

function Dashboard() {
  const [locations, setLocations] = useState(null);
  const [error, setError] = useState(null);
  const [unit, setUnit] = useState(localStorage.getItem("toggle") === "false");
  const [hide, setHide] = useState(false);
  const [geolocation, setGeolocation] = useState(
    localStorage.getItem("allow") === "true",
  );

  const allowLocation = (isAllowed) => {
    localStorage.setItem("allow", JSON.stringify(isAllowed));
    setGeolocation(isAllowed);
  };

  const fetchLocations = () => {
    navigator.geolocation.getCurrentPosition((position) => {
      getLocations(
        setLocations,
        setError,
        position.coords.latitude,
        position.coords.longitude,
      );
    });
  };

  useEffect(() => {
    // Check permissions.query exists
    let isComponentMounted = true;
    if (isComponentMounted) {
      navigator.permissions.query({ name: "geolocation" }) &&
        navigator.permissions.query({ name: "geolocation" }).then((result) => {
          if (result.state === "granted") {
            fetchLocations();
            allowLocation(true);
          }
          if (result.state === "prompt") {
            fetchLocations();
          }
          if (result.state === "denied") {
            getLocations(setLocations, setError);
            allowLocation(false);
          }
        });
    }
    return () => {
      isComponentMounted = false;
    };
  }, []);

  // On permission change, update location data
  useEffect(() => {
    navigator.permissions.query({ name: "geolocation" }) &&
      navigator.permissions.query({ name: "geolocation" }).then((result) => {
        result.onchange = function () {
          if (result.state === "granted") {
            fetchLocations();
            allowLocation(true);
          }
          if (result.state === "denied") {
            getLocations(setLocations, setError);
            allowLocation(false);
          }
        };
      });
  });

  if (error) {
    return (
      <span style={{ padding: "1em" }}>
        There was a problem loading the weather forecast. Please allow location
        services, try and refresh the page or restart your internet connection.
      </span>
    );
  }

  return (
    <div>
      <ToggleSwitch
        data-testid="toggle-switch"
        changeUnit={() => changeUnit(unit, setUnit)}
        unit={unit}
      />
      <Layout>
        <h1 style={{ padding: "1em 0" }}>Dashboard</h1>
        <CardList locations={locations} unit={unit} />
        <LocationSearch setLocations={setLocations} setError={setError} />
      </Layout>
      {!geolocation && !hide && (
        <PopupContainer>
          <Popup>
            <span>
              Please allow location services in your browser settings to view
              local weather information
            </span>
            <button onClick={() => setHide(true)}>ok</button>
          </Popup>
        </PopupContainer>
      )}
    </div>
  );
}

export default Dashboard;
