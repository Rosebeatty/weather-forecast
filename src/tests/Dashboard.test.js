import React from "react";
import { render } from "@testing-library/react";
import Dashboard from "pages/Dashboard";
import { MemoryRouter, useNavigate } from "react-router-dom";

// Mock useNavigate
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: jest.fn(),
}));

describe("<Dashboard />", () => {
  const mockNavigate = jest.fn();

  beforeEach(() => {
    jest.mocked(useNavigate).mockReturnValue(mockNavigate);

    localStorage.setItem("toggle", "false");
    render(
      <MemoryRouter>
        <Dashboard />
      </MemoryRouter>
    );
  });

  it("renders without crashing", () => {
    const { asFragment } = render(<Dashboard />);
    expect(asFragment()).toMatchSnapshot();
  });
});
