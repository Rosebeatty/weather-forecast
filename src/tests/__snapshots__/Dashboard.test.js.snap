// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> renders without crashing 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Dashboard
    </h1>
    <div
      class="sc-jXbVAB ENxqO"
      for="toggle"
    >
      <div
        class="react-toggle react-toggle--checked"
      >
        <div
          class="react-toggle-track"
        >
          <div
            class="react-toggle-track-check"
          >
            <p
              class="sc-dhKdPU iHtkx"
            >
              °C
            </p>
          </div>
          <div
            class="react-toggle-track-x"
          >
            <p
              class="sc-dhKdPU iHtkx"
            >
              °F
            </p>
          </div>
        </div>
        <div
          class="react-toggle-thumb"
        />
        <input
          aria-label="No label"
          checked=""
          class="react-toggle-screenreader-only"
          data-testid="toggle"
          id="toggle"
          type="checkbox"
        />
      </div>
    </div>
    <div
      class="sc-eDPFhE eJcoVx"
    >
      <div
        class="sc-gsFSjX bKuoSs"
      >
        <div
          class="sc-imWZod kRiiPg"
        >
          <div
            aria-busy="true"
            aria-label="tail-spin-loading"
            class="sc-aYaIB nSnSW"
            data-testid="tail-spin-loading"
            role="progressbar"
          >
            <svg
              data-testid="tail-spin-svg"
              height="80"
              viewBox="0 0 38 38"
              width="80"
              xmlns="http://www.w3.org/2000/svg"
            >
              <defs>
                <lineargradient
                  id="a"
                  x1="8.042%"
                  x2="65.682%"
                  y1="0%"
                  y2="23.865%"
                >
                  <stop
                    offset="0%"
                    stop-color="purple"
                    stop-opacity="0"
                  />
                  <stop
                    offset="63.146%"
                    stop-color="purple"
                    stop-opacity=".631"
                  />
                  <stop
                    offset="100%"
                    stop-color="purple"
                  />
                </lineargradient>
              </defs>
              <g
                fill="none"
                fill-rule="evenodd"
              >
                <g
                  transform="translate(1 1)"
                >
                  <path
                    d="M36 18c0-9.94-8.06-18-18-18"
                    id="Oval-2"
                    stroke="purple"
                    stroke-width="2"
                  >
                    <animatetransform
                      attributeName="transform"
                      dur="0.9s"
                      from="0 18 18"
                      repeatCount="indefinite"
                      to="360 18 18"
                      type="rotate"
                    />
                  </path>
                  <circle
                    cx="36"
                    cy="18"
                    fill="#fff"
                    r="1"
                  >
                    <animatetransform
                      attributeName="transform"
                      dur="0.9s"
                      from="0 18 18"
                      repeatCount="indefinite"
                      to="360 18 18"
                      type="rotate"
                    />
                  </circle>
                </g>
              </g>
            </svg>
          </div>
        </div>
      </div>
      <div
        class="sc-kpDprT oFLHV"
      >
        <div
          class="sc-cwHqhk kMCbFI"
        >
          <label
            class="sc-jlZhRR kiUyVY"
            for="date"
          >
            Date:
          </label>
          <div>
            <div
              class="react-datepicker-wrapper"
            >
              <div
                class="react-datepicker__input-container"
              >
                <input
                  class="sc-dAlxHm oJzHo"
                  placeholder="Pick a Date"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="sc-cwHqhk kMCbFI"
          >
            <label
              class="sc-jlZhRR kiUyVY"
              for="location"
            >
              Location:
            </label>
            <div
              style="display: flex; flex-direction: column;"
            >
              <input
                class="sc-dAlxHm oJzHo"
                id="location"
                placeholder="City, Town or Zip Code"
                type="text"
                value=""
              />
            </div>
          </div>
          <p
            style="font-weight: bold; padding: 1em 0px;"
          >
            OR
          </p>
          <div
            class="sc-cwHqhk kMCbFI"
          >
            <label
              class="sc-jlZhRR kiUyVY"
              for="latitude"
            >
              Latitude: 
              <span
                style="font-size: 10px;"
              >
                (-90 to 90)
              </span>
            </label>
            <input
              class="sc-dAlxHm oJzHo"
              id="latitude"
              max="90"
              min="-90"
              placeholder="-5"
              style="width: 100%;"
              type="number"
              value=""
            />
            <label
              class="sc-jlZhRR kiUyVY"
              for="longitude"
            >
              Longitude: 
              <span
                style="font-size: 10px;"
              >
                (-180 to 180)
              </span>
            </label>
            <input
              class="sc-dAlxHm oJzHo"
              id="longitude"
              max="180"
              min="-180"
              placeholder="10"
              style="width: 100%;"
              type="number"
              value=""
            />
          </div>
        </div>
        <div>
          <button
            class="sc-jEABJj ePbAxF"
            disabled=""
          >
            Search
          </button>
        </div>
      </div>
    </div>
    <div
      class="sc-eldOKa bdUXFi"
    >
      <div
        class="sc-fPXMhL cqsCCa"
      >
        <span>
          Please allow location services in your browser settings to view local weather information
        </span>
        <button>
          ok
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
